### DISCUSSION TITLE
How to prevent multiple logins from same user?

### DISCUSSION THREAD

**Discussion by Oggy4th** [pocketbase/pocketbase#2042] [Q&A]
I have 2 frontends. Windows desktop app in **Flutter** and **Svelte** web.

Both frontend using `pb.collection('users').authWithPassword()`.

Also I have backend using pocketbase as framework.

I wanna prevent multiple logins from same user.

Should I use golang library named **gorilla/sessions**.

What if internet connection or electricity cut off, how to delete session data? and how to login user again?

**Comment by ganigeorgiev** [comment]
I'm not sure that I understand your use case.

Could you elaborate a little more what are you trying to do?

Why do you want to prevent users being logged in from different apps?

If you just want to sync the user data in both apps, you can subscribe to the user realtime events, eg:

```js
pb.collection("users").subscribe(pb.authStore.model.id, (e) => {
    console.log(e);
});
```

**Reply by Oggy4th** [reply]
Hello, there @ganigeorgiev 

Thanks for your reply

**Sorry my bad.**

Users will use Windows desktop app in Flutter. `pb.collection('users').authWithPassword()`.

Svelte web is admin panel. `pb.collection('employees').authWithPassword()`.

Users can't login admin panel. Also admins can't login desktop app.

I wanna prevent to login someone who know my account name and password after I've already login my account.

**Reply by ganigeorgiev** [reply]
PocketBase web APIs are stateless, aka. there are no sessions and a client request is considered authenticated as long as there is a valid `Authorization: TOKEN` header sent with the request.

I'm still not sure that I understand what you are trying to do and why.

Even if you implement an additional session layer and prevent other users to logged in, what should be the behavior if the malicious user establishes a session before you?
Based on your requirements, it seems that while the malicious user is logged in, you'll be locked out of your own account.

A better approach probably would be to have a "Known devices" list and send immediately an email if someone sign-in with your account for the first time.
Other option would be to have 2FA (_one time email code for admins may be added in the future_).
You can also implement an IP whitelist/blacklist filter either inside the Go app or on a reverse-proxy level.

**Reply by Oggy4th** [reply]
There have public computers. Any one can login by their account using desktop app.
I am doing some compute with user data. If I allow multiple login, my compute will work wrong.
That's why I only wanna prevent multiple logins.

what should be the behavior if the malicious user establishes a session before you?
They can login. it doesn't matter.

**Comment by ghost** [comment]
I'm trying to implement something like this also, I would like to prevent simultaneous login of the same user on any mobile device. The reason I don't want this is that I'm retrieving location of the user to display on the dashboard, but if they login from different device simultaneously data's will override each other. I'm thinking about storing status on user collection to identify if it's already logged in or not and block any other login requests but this might fail due to status field not updating 100% of the time.

**Comment by ganigeorgiev** [comment]
@temzints, @Oggy4th 

I don't think blocking a user login is the correct solution for what you are trying to do, but I don't have the full picture and I still don't think that  I fully understand your use cases so I could be wrong.

If you want to have only a single valid user auth token at a time, the easiest approach would be to invalidate all previously issued tokens on every login, aka. always the last login "win".
There is no direct endpoint for invalidating tokens, but you can trigger the user tokens invalidation by setting the password to the same value after login.

```js
async function loginInvalidatingOldTokens(email, pass) {
    await pb.collection("users").authWithPassword(email, pass)

    await pb.collection("users").update(pb.authStore.model.id, {
        oldPassword:     pass,
        password:        pass,
        passwordConfirm: pass,
    });
    
    await pb.collection("users").authWithPassword(email, pass)
}

await loginInvalidatingOldTokens("test@example.com", "1234567890");
```

If you want to do the above on the server-side, you can attach to the `OnRecordAuthRequest` hook and refresh the user's `tokenKey`:

```go
app.OnRecordAuthRequest().Add(func(e *core.RecordAuthEvent) error {
	// generate and set a new random token key
	e.Record.RefreshTokenKey()

	if err := app.Dao().SaveRecord(e.Record); err != nil {
		return err
	}

	token, err := tokens.NewRecordAuthToken(app, e.Record)
	if err != nil {
		return err
	}

	e.Token = token

	return nil
})
```

(_I haven't tested any of the above examples, so you may have to adjust them based on your use cases_).

**Reply by tonnax** [reply]
Hi @ganigeorgiev may I know what is the latest syntax for the Go **server-side** case? Dao, Add, and tokens are returning syntax error in my IDE. I'm using the Pocketbase v0.25.0.


is this the correct one? 
```Go
        app.OnRecordAuthRequest("users").BindFunc(func(e *core.RecordAuthRequestEvent) error {
		if e.Request != nil && strings.Contains(e.Request.URL.Path, "/auth-refresh") {
			return e.Next()
		}
		e.Record.RefreshTokenKey()
		if err := app.Save(e.Record); err != nil {
			return apis.NewBadRequestError("Failed to refresh token key", err)
		}
		return e.Next()
	})
	
```

and if yes, in my client side, do I need to call "/auth-refresh" after my  "/auth-with-otp"  response success?

**Reply by ganigeorgiev** [reply]
@tonnax Please don't hijack old discussions. The above code samples are for PocketBase <= v0.22.x. Open a new one and elaborate clearly on your use case and what actually you are trying to do.

**Reply by ganigeorgiev** [reply]
@tonnax Additionally please read also https://github.com/pocketbase/pocketbase/discussions/2042#discussioncomment-5275124.

**Comment by airnh** [comment]
I think this is helpful as my application assumes 1 user per device and I wish to prevent multiple users share a credential. If the second user logs in with the credential, the session in the previous device with the same account will be logged out.

So, whenever a user logges in, the default PB token is sent to an extra backend and update the a session token (added field) and validates or so..

with chatGPT and the psuedo code...
````
const express = require('express');
const httpProxy = require('http-proxy');

const app = express();
const proxy = httpProxy.createProxyServer();

// Define the middleware function to check cookie and endpoint validity
const validatorMiddleware = (req, res, next) => {

  // Check if the endpoint URL contains a "Cold Sign In,  Refresh, or after PB-sign in", then do nothing and forward.
  if (req.path.includes('/auth-refresh', '/auth-with-oauth2', 'auth-with-password', 'my_endpoint_to_create/update_new_token')) {
    next()
  }

  // Other than this, the request should include  the session cookie with the token
  if (!req.headers.cookie || req.headers.cookie !== 'mycookie' || !req.headers.cookie.token ) {
    return res.status(401).send('Invalid Cookie');
  }

  // if token presents
  try{
    var res = await pb.collection("users").getFirstListItem(`session_token=='${token}'`)
  }catch(e){
    res.setheader -> delete cookie
    return res.status(401).send('session expired');
  }  
  
  // if session token is good, 
  next();
};

// Use the middleware function for all routes
app.use(validatorMiddleware, (req, res) => {
  // Forward the request to the target server
  proxy.web(req, res, { target: 'http://targetserver.com' });
});

// Start the server
app.listen(3000, () => {
  console.log('Proxy server listening on port 3000');
});
```

**Reply by ganigeorgiev** [reply]
@airnh You don't need session cookies if all you want is to invalidate the previously issued auth tokens.
See the code samples in https://github.com/pocketbase/pocketbase/discussions/2042#discussioncomment-5275523.

**Reply by airnh** [reply]
OH, the old and the new password can be the same to still invalidate?

**Reply by ganigeorgiev** [reply]
Yes, any valid submitted password value should reset the user's token key.

**Reply by airnh** [reply]
saved my life from my assumption!

**Reply by airnh** [reply]
@ganigeorgiev I have tested for some cases. 

1. when the user has an email and password, no problem. 
2. when the user has social login, you can still create a random string and re-login with id and password. ok. It looks I can't use authWithOAuth2 twice.
3. when the user has both email&pw and social login and the user logs in with id and password. ok
4. when the user has both email&pw and social login and the user logs in with social account.  Not sure. If the user has a its own password, there's no way I can invalidate with the previous trick. (as even admin cannot see the user's password)

So for now, I should just enforce the user have either social account or email. (Assuming my understanding of PB is correct....)

---

But, can I? If the user already has an email and pw, can I stop the user from logging in with the social account? As the redirect url doesn't tell anything about the user's email as far as I understand.

**Reply by ganigeorgiev** [reply]
@airnh If you have OAuth2 login enabled, you can apply the server-side approach using the `OnRecordAuthRequest` hook since it handles all auth scenarios.

**Reply by airnh** [reply]
Please bear with me - to implement the serverside, do I need to run it as a go app or can I achive it with the prebuilt app as well? Please le me know any relevant links if applicable. Thanks in advance.

**Reply by ganigeorgiev** [reply]
@airnh You have to build the executable on your own.
More information you can find in https://pocketbase.io/docs/use-as-framework/.
The prebuilt executable main file could be found in - https://github.com/pocketbase/pocketbase/blob/master/examples/base/main.go.

**Reply by airnh** [reply]
The serverside hook is smooth!

**Reply by airnh** [reply]
@ganigeorgiev Is it possible to bypass the token invalidation if it is "auth-refresh". I tried other hooks but their event types don't seem to have e.Token

I have an extra backend that needs to check user's token is valid to authorize other services. FYI, I'm using fetch instead of SDK as the SDK on the server side after auth-refresh will overide admin token anyway.

The below is ad hoc. I wonder if there could be better way wihtout using extra field "sys". If I do not add the field "sys" from the admin panel, it still works while the field value is not returned in the response (which is good). Does it have any potential issue to break the collection structure or so?
```
app.OnRecordAuthRequest().Add(func(e *core.RecordAuthEvent) error {
		log.Println("on record atuh request")

		// generate and set a new random token key
		e.Record.RefreshTokenKey()
		log.Println(e.Record.Get("sys"))
		if e.Record.Get("sys") == true {
			log.Println("bypass")
			e.Record.Set("sys", false)
			return nil
		}
		log.Println("new token")

		if err := app.Dao().SaveRecord(e.Record); err != nil {
			return err
		}

		token, err := tokens.NewRecordAuthToken(app, e.Record)
		if err != nil {
			return err
		}

		e.Record.TokenKey()
		e.Token = token

		log.Println(e.Record.TokenKey())
		log.Println(e.Token)

		return nil
	})

	app.OnRecordBeforeAuthRefreshRequest().Add(func(e *core.RecordAuthRefreshEvent) error {
		log.Println("OnRecordBeforeAuthRefreshRequest")
		log.Println("setting for bypass")
		e.Record.Set("sys", true)
		return nil
	})
```

**Reply by ganigeorgiev** [reply]
> Is it possible to bypass the token invalidation if it is "auth-refresh".

You can check the `e.HttpContext` route path inside the hook and ignore further execution, something like:

```go
if strings.Contains(e.HttpContext.Path(), "/auth-refresh") {
	return nil
}
```

---

> I'm using fetch instead of SDK as the SDK on the server side after auth-refresh will overide admin token anyway.

That's OK. Note that there is also nothing wrong in creating a one-off new SDK instances (aka. `new PocketBase(...)`).

--- 

> If I do not add the field "sys" from the admin panel, it still works while the field value is not returned in the response (which is good). Does it have any potential issue to break the collection structure or so?

No, it is fine (_but I recommend the path check approach above_). Unknown record data is returned in the response only if you enable it with `e.Record.WithUnkownData(true)`.

**Reply by airnh** [reply]
@ganigeorgiev This is getting a bit funny. 

In short, my question might have been murky for the entire time. Is there a way I can only check if a requested token is valid? I was expecting something like Firebase's Verify ID Tokens: [FB](https://firebase.google.com/docs/auth/admin/verify-id-tokens)

The desired behavior is that

For all new logins: invalidate all previous tokens and return a new token. Your example works!
For auth-refresh, keep the same token and only check the validity (expiration) of the token.

It seems like the code for auth-refresh just keeps generating new tokens.. obviously. 

if strings.Contains(e.HttpContext.Path(), "/auth-refresh") {
	return nil  // this was just going back to the default behavior, generating multiple valid tokens and return.
}

**Reply by ganigeorgiev** [reply]
@airnh 

> Is there a way I can only check if a requested token is valid?

There is no dedicated endpoint for that. You have to send `POST /api/collections/users/auth-refresh` request. On invalid or expired token it returns an error, otherwise - new token and data.

> For auth-refresh, keep the same token and only check the validity (expiration) of the token.

But why it matters whether a new token is generated?
The new token will be still related to the same user's `tokenKey`, just the `exp` claim is different. If you have applied the server-side check and the `auth-refresh` ignore case, whenever the user authenticate from a different device (and via non `/auth-refresh` request) all previously issued tokens will be invalidated anyway.

**Reply by airnh** [reply]
> But why it matters whether a new token is generated? The new token will be still related to the same user's `tokenKey`, just the `exp` claim is different. If you have applied the server-side check and the `auth-refresh` ignore case, whenever the user authenticate from a different device (and via non `/auth-refresh` request) all previously issued tokens will be invalidated anyway.

You are right! I was so fixed that if the old and new tokens don't match, that's the invalidated case only.

Thanks, alot!!! @ganigeorgiev
